### PR TITLE
HDE-14 audit logging support

### PR DESCRIPTION
**Pull request by emreathelix**

### CODE REVIEW

File: providers/structured/provider.go [hunk 1]
```diff
@@ -28,6 +29,8 @@ type RecordEventCallArgs struct {
 }
 
 type StructuredOutputLogProvider struct {
+	providers.LogProvider
```
Comment (jpecknerhelix) [inline providers/structured/provider.go]: This is actually the first time I've looked at the way we do log chaining, so let me know if this is incorrect, but does the next provider have to be explicitly called in all the various log methods below? (Not seeing it being invoked elsewhere.) I.e., in `func (p *StructuredOutputLogProvider) Error(...)`, you'd call 

    if p.nextProvider != nil {
        p.nextProvider.Error(ctx, report, args...)
    }

And same for the other six methods defined in `LogProvider`.

File: providers/structured/provider.go [hunk 2]
```diff
@@ -47,16 +50,25 @@ func (p StructuredOutputLogProvider) GetRecordEventCalls() []RecordEventCallArgs
...
 	}
 
-	return &StructuredOutputLogProvider{
+	lp := &StructuredOutputLogProvider{
 		rawLogCalls:      rawLogCalls,
 		recordCalls:      []RecordCallArgs{},
 		recordEventCalls: []RecordEventCallArgs{},
 	}
+
+	if nextProvider != nil {
```
Comment (jpecknerhelix) [inline providers/structured/provider.go]: Possible to test the behavior of this block in the unit tests? The rest of this method is already covered, so you'd just need to add a test for `nextProvider == nil`/another for `nextProvider != nil`.

File: VERSION [hunk 3]
```diff
@@ -1 +1 @@
-1.2.1
+1.2.2
```
[the remark this answers is not in the document]
Comment (emreathelix) [reply VERSION]: Oops sorry about that

File: providers/structured/provider.go (hunk 1 above)
Comment (emreathelix) [reply providers/structured/provider.go]: I think you're right. Otherwise, it wouldn't really be chaining I guess.

Comment (jpecknerhelix) [reply providers/structured/provider.go]: Sounds good- feel free to check that with Chris if there's any ambiguity, since there could be something I'm missing.

File: VERSION (hunk 3 above)
[the remark this answers is not in the document]
Comment (jpecknerhelix) [reply VERSION]: No worries, easy to fix!

File: providers/structured/provider_test.go [hunk 4]
```diff
@@ -171,4 +175,66 @@ var _ = Describe("bufferedLogProvider", func() {
...
+		// put some events
+		lp.RecordEvent(contextLogger, "Event1", log.Metrics{
+			"Metric 1": "Value 1",
+			"Metric 2": "Value 2",
+		})
+		lp.RecordEvent(contextLogger, "Event2", log.Metrics{
+			"Metric 3": "Value 3",
+			"Metric 4": "Value 4",
+		})
+
```
Comment (jpecknerhelix) [inline providers/structured/provider_test.go]: Think you should also test that calling `lp.Wait()` causes `lp.nextProvider.Wait()` to be called. Looks like you could use dummy/provider as essentially a mock for `nextProvider`, to accomplish that verification.

**Comment by emreathelix** [comment]
Thanks for all the help @jpecknerhelix

File: providers/structured/provider_test.go [hunk 5]
```diff
@@ -186,6 +186,17 @@ var _ = Describe("bufferedLogProvider", func() {
 		Ω(lp.LogProvider).Should(Equal(chaining.LogProvider(dummyProvider)))
 	})
 
+	It("Should call wait on the next provider", func() {
+		ws := new(dummy.WaitState)
```
Comment (jpecknerhelix) [inline providers/structured/provider_test.go]: Nit: add `Ω(ws.Get()).Should(BeFalse())` after this line as a sanity check

**Comment by jpecknerhelix** [comment]
No problem! Btw, are the unit tests for the `reported_at` suite failing on your machine too? I'm seeing this in the last few master commits, so it wasn't introduced in this branch. So IMO it's not a blocker for this PR.

**Comment by jpecknerhelix** [comment]
^ @emreathelix, CC: @chriswhelix

**Comment by emreathelix** [comment]
@jpecknerhelix All tests are passing for me both on master and this branch. Make sure you've run `glide up`, `glide install`

**Comment by jpecknerhelix** [comment]
Weird, I'm actually seeing _more_ errors after running `glide up` and `glide install`.

**Comment by jpecknerhelix** [comment]
Created a ticket in JIRA for the failing test issue.

File: log/log.go [hunk 6]
```diff
@@ -24,6 +24,10 @@ func SetDefaultProvider(provider providers.LogProvider) {
 	defaultProvider = provider
 }
 
+func GetDefaultProvider() providers.LogProvider {
```
Comment (chriswhelix) [inline log/log.go]: Idiomatic name in Go should be just DefaultProvider().

File: providers/structured/provider.go [hunk 7]
```diff
@@ -47,16 +50,25 @@ func (p StructuredOutputLogProvider) GetRecordEventCalls() []RecordEventCallArgs
...
+		lp.LogProvider = chaining.LogProvider(nextProvider)
```
Comment (chriswhelix) [inline providers/structured/provider.go]: What you want to do is here always set lp.LogProvider to the chaining provider, even if nextProvider is nil; then you don't need to bother with all the nil-checking in individual methods, because the chaining provider does it for you (that's its main job).

File: providers/structured/provider.go [hunk 8]
```diff
@@ -104,6 +132,14 @@ func (p *StructuredOutputLogProvider) RecordEvent(ctx context.Context, eventName
...
+		p.LogProvider.RecordEvent(ctx, eventName, metrics)
+	}
 }
 
-func (p *StructuredOutputLogProvider) Wait() {}
+func (p *StructuredOutputLogProvider) Wait() {
+	if p.LogProvider != nil {
+		p.LogProvider.Wait()
+	}
+}
```
Comment (chriswhelix) [inline providers/structured/provider.go]: You don't need to implement this at all; the chaining provider will give you an empty default implementation.

File: providers/structured/provider.go [hunk 9]
```diff
@@ -47,16 +50,25 @@ func (p StructuredOutputLogProvider) GetRecordEventCalls() []RecordEventCallArgs
 
 // NewStructuredOutputLogProvider returns a LogProvider which records all calls made to it.
 func NewStructuredOutputLogProvider() *StructuredOutputLogProvider {
+	return LogProvider(nil).(*StructuredOutputLogProvider)
+}
+
+func LogProvider(nextProvider providers.LogProvider) providers.LogProvider {
```
Comment (chriswhelix) [inline providers/structured/provider.go]: Why do we need two different constructors here, with different signatures?

Comment (emreathelix) [reply providers/structured/provider.go]: Because `NewStructuredOutputLogProvider()` is already being used in other services. I didn't wanna break any existing functionality.

File: providers/structured/provider.go (hunk 7 above)
Comment (emreathelix) [reply providers/structured/provider.go]: Ah ok.. Didn't check the internals of the chaining provider.

File: providers/structured/provider.go (hunk 9 above)
Comment (chriswhelix) [reply providers/structured/provider.go]: Hmm, OK. Let's mark the old one deprecated using the Go convention (see: https://rakyll.org/deprecated/), and have this also return *StructuredOutputLogProvider to avoid awkward casting of its result.
